### ISSUE TITLE
Create AWS::Serverless::GatewayResponse resource

### ISSUE THREAD

**Issue by sthulb** [awslabs/serverless-application-model#815]
There's a number of issues raised by customers regarding the mix of `AWS::Serverless::Api` and `AWS::ApiGateway::GatewayResponse`

We should do better and create a resource to ensure this works. From my understanding this is a result of merge behaviour and quirks in CFN.

#623

**Comment by carlnordenfelt** [comment]
Would it make sense, to extend the `AWS::Severless:Api` resource with gateway responses to avoid adding another Serverless resource?

Example:
    ExampleAPI:
      Type: 'AWS::Serverless::Api'
      Properties:
        GatewayResponses:
          - ResponseParameters:
               gatewayresponse.header.Access-Control-Allow-Origin: "'*'"
             ResponseType: 'DEFAULT_4XX'

I think the above would be very useful and a 'simple' solution to the problem.

**Comment by brettstack** [comment]
@carlnordenfelt is correct, we would need to extend the `AWS::Serverless::Api` with a `GatewayResponses` parameter instead of creating a new resource. In this case we would probably just pass through anything defined on `GatewayResponses` and apply it as `x-amazon-apigateway-gateway-responses` in the Swagger body.

Note we are talking more about the "Merge" feature I alluded to in #623 which will effectively do this pass through.

**Comment by brettstack** [comment]
For anyone who wants to have a go at implementing this, here are some relevant places in code:

1. https://github.com/awslabs/serverless-application-model/blob/master/samtranslator/model/apigateway.py#L11
1. https://github.com/awslabs/serverless-application-model/blob/master/samtranslator/model/api/api_generator.py#L52
1. Apply the value to swagger https://github.com/awslabs/serverless-application-model/blob/master/samtranslator/swagger/swagger.py#L33

**Comment by carlnordenfelt** [comment]
If only it wasn't python :)

**Comment by brettstack** [comment]
@carlnordenfelt curious which language you would prefer.

**Comment by carlnordenfelt** [comment]
@brettstack feels like we are hijacking the issue a bit but my go to is nodejs or even Java. Not too familiar with the ins and outs of python.

**Comment by Buffer0x7cd** [comment]
@brettstack I would love to work on the above issue.

**Comment by keetonian** [comment]
@Buffer0x7cd Absolutely! Go ahead and give it a try, and reach out if you have any questions!

**Comment by chrisoverzero** [comment]
@Buffer0x7cd I'm also working on this -- I hope we're not duplicating each other's work too much! Let's sync up if you've already started.

In my implementation, I've broken it down a little more than it was in OP. For example, to recreate (kinda) my use-case from #623:

```yaml
ExplicitApi:
  Type: AWS::Serverless::Api
  Properties:
    StageName: Prod
    GatewayResponses:
      UNAUTHORIZED:
        StatusCode: 401
        ResponseParameters:
          Headers:
            Access-Control-Expose-Headers: "'WWW-Authenticate'"
            Access-Control-Allow-Origin: "'*'"
            WWW-Authenticate: >-
              'Bearer realm="admin"'
```

(I just grabbed this from one of my test cases -- please don't think it's fully representative.)

I'm a hair over 95% coverage, but I'm not sure I've covered all the negative cases I want to. Once the codey part is done, I'll open an MR. I think I'll need more help with the spec language than with the Python.

**Comment by chrisoverzero** [comment]
This is now in #841.

Hey, while I'm here, is the translator meant to be runnable from the `bin` directory? I had to move it up one to the project root before it would run.

**Comment by keetonian** [comment]
@chrisoverzero I do that too haha. It has a few other bugs as well, it needs to be updated.

**Comment by keetonian** [comment]
Released with [SAM v1.11.0](https://github.com/awslabs/serverless-application-model/releases/tag/v1.11.0)!

**Comment by FredrikMeyer** [comment]
There should be some examples of this somewhere. I had to dig through the PR mentioned above to find the syntax.
